### PR TITLE
feat(cli): add mcp prompt support via slash commands

### PR DESCRIPTION
**Pull request by kalvinnchau**
## update `Prompt` structs for align with MCP spec, add basic prompt commands to mcp-client

* [Prompts](https://modelcontextprotocol.io/docs/concepts/prompts#prompt-structure) `description` and `arguments` are optional fields, use optional in the structs
   * similarly `PromptArguments` `description` and `required` are optional too
* add `list_prompts` and `get_prompt` method to the `McpClientTrait`
* add implementation of `list_prompts` and `get_prompt` to call `prompts/list` and `prompts/get` respectively
* update `stdio` and `sse` clients to handle `JsonRpcMessage::Error` responses to send back to the user instead of doing nothing

### CODE REVIEW

File: crates/goose-mcp/src/developer/mod.rs [hunk 1]
```diff
@@ -70,9 +70,9 @@ pub fn load_prompt_files() -> HashMap<String, Prompt> {
                 description: arg.description,
                 required: arg.required,
             })
-            .collect();
+            .collect::<Vec<PromptArgument>>();
```
Comment (salman1993) [inline crates/goose-mcp/src/developer/mod.rs:75]: unrelated to this PR: i don't think we need this mapping  - https://github.com/block/goose/blob/main/crates/goose-mcp/src/developer/prompts/unit_test.json

it'd be easier to stay closer to the MCP prompt in the json to avoid this mapping (description instead of template)

Comment (salman1993) [reply crates/goose-mcp/src/developer/mod.rs:73]: i'd call it `description` instead of template for example and stay closer to the hierarchy in MCP spec

Comment (salman1993) [reply crates/goose-mcp/src/developer/mod.rs:73]: nvm maybe we do need the template cause i don't see any way to replace the arguments - https://spec.modelcontextprotocol.io/specification/2024-11-05/server/prompts/#text-content

**Review by salman1993 — APPROVED** [review]

**Comment by salman1993** [comment]
nit: you can add a prompt in the mcp-server example here: https://github.com/modelcontextprotocol/rust-sdk/blob/main/crates/mcp-server/src/main.rs. that'll make it easier to test from this mcp-client example: https://github.com/modelcontextprotocol/rust-sdk/blob/main/crates/mcp-client/examples/stdio_integration.rs

**Comment by kalvinnchau** [comment]
Moved this back into a draft PR for now, since we're in progress of moving the mcp stuff to the [rust-sdk](https://github.com/modelcontextprotocol/rust-sdk), I'm keeping the commits in sync manually for now but I'll hold off on diverging for nwo.

**Comment by kalvinnchau** [comment]
Added some basic cli based prompting scaffolding, mainly to list prompts and get the information about them.

Calling the prompt is not yet implemented.

Ex:
```bash
( O)> /prompts

 mcp_test
  - no_args_prompt
  - one_arg_prompt
 developer
  - unit_test

( O)> /prompt unit_test --info

 Extension: developer
Prompt: unit_test

 Generate or update unit tests for a given source code file.

The source code file is provided in {source_code}.
Please update the existing tests, ensure they are passing, and add any new tests as needed.

The test suite should:
- Follow language-specific test naming conventions for {language}
- Include all necessary imports and annotations
- Thoroughly test the specified functionality
- Ensure tests are passing before completion
- Handle edge cases and error conditions
- Use clear test names that reflect what is being tested

 Arguments:
  source_code (required) The source code file content to be tested
  language (required) The programming language of the source code

( O)>
```

With prompt execution support:
```bash
/prompt unit_test source_code="~/stage/unit_test/file.py" language=python
# Unit Test Generation and Update

I'll help you generate or update unit tests for the source code file at `~/stage/unit_test/file.py`. First, let me check if this file exists and examine its c
ontents to understand what we need to test.

─── shell | developer ──────────────────────────
command: ls -la ~stage/unit_test/
...

Now, let's run the existing tests to see if they pass:

─── shell | developer ──────────────────────────
command: cd /Users/kalvin/stage/unit_test/ && python -m unittest test_file.py


.....
----------------------------------------------------------------------
Ran 5 tests in 0.005s

OK

Great! The existing tests are already passing. Let's analyze the current test coverage and see if we need to add any additional tests:

...

However, we can enhance the test suite with a few additional tests:

1. Test reading a file with special characters
2. Test reading a file with different encodings
3. Test reading a large file (performance test)

Let's update the test file with these additional tests:

─── text_editor | developer ──────────────────────────
path: ~/stage/unit_test/test_file.py
```

**Comment by kalvinnchau** [comment]
https://github.com/user-attachments/assets/f0d7d511-61ce-4b80-8799-4a0305baa842

File: crates/goose/src/message.rs [hunk 2]
```diff
@@ -156,6 +158,37 @@ impl From<Content> for MessageContent {
     }
 }
 
+impl From<PromptMessage> for Message {
+    fn from(prompt_message: PromptMessage) -> Self {
+        // Create a new message with the appropriate role
+        let message = match prompt_message.role {
+            PromptMessageRole::User => Message::user(),
+            PromptMessageRole::Assistant => Message::assistant(),
```
Comment (salman1993) [inline crates/goose/src/message.rs:166]: the code herer looks fine.

in the CLI and server:
if we invoked a PromptMessage that starts with assistant role - i _think_ that would error because the CLI assumes its user msg and we also require alternative between user & assistant. can you check this?

File: crates/goose-cli/src/session/mod.rs [hunk 3]
```diff
@@ -104,6 +108,40 @@ impl Session {
...
+                    arguments: prompt.arguments.clone(),
+                    extension: Some(extension),
+                }));
+            }
+        }
+
+        Ok(None)
+    }
+
+    pub async fn get_prompt(&mut self, name: &str, arguments: Value) -> Result<Vec<PromptMessage>> {
```
Comment (salman1993) [inline crates/goose-cli/src/session/mod.rs:140]: if we invoked a PromptMessage that starts with assistant role - i think that would error because the CLI assumes its user msg and we also need to alternate between user & assistant. can you check this? if so, we might wanna enforce some checks.

File: crates/goose-cli/src/session/mod.rs [hunk 4]
```diff
@@ -179,6 +217,42 @@ impl Session {
...
+                                output::render_error(&format!("Prompt '{}' not found", opts.name))
+                            }
+                        }
+                    } else {
+                        // Convert the arguments HashMap to a Value
+                        let arguments = serde_json::to_value(opts.arguments)
+                            .map_err(|e| anyhow::anyhow!("Failed to serialize arguments: {}", e))?;
+
+                        match self.get_prompt(&opts.name, arguments).await {
+                            Ok(messages) => {
```
Comment (salman1993) [inline crates/goose-cli/src/session/mod.rs:243]: do you think its useful to render the expanded prompt here?

Comment (kalvinnchau) [reply crates/goose-cli/src/session/mod.rs:243]: oh yeah, we should - the user might be confused if its just `/prompt ...` and things start happening without showing the expanded prompt.

File: crates/goose-cli/src/session/mod.rs (hunk 3 above)
Comment (kalvinnchau) [reply crates/goose-cli/src/session/mod.rs:140]: added enforcement for user first, and that the messages alternate user & assistant
